rust/lance-encoding/src/encodings/physical/dictionary.rs

```diff
@@ -0,0 +1,410 @@
+        let null_buffer = string_array
+            .nulls()
+            .map(|n| BytesMut::from(n.buffer().as_slice()))
+            .unwrap_or_else(BytesMut::new);
+
+        let offsets_buffer = BytesMut::from(string_array.offsets().inner().inner().as_slice());
+
+        // Empty buffer for nulls of bytes
+        let empty_buffer = BytesMut::new();
+        // let bytes_buffer = string_array.values().as_slice().into();
+        let bytes_buffer = BytesMut::from_iter(string_array.values().iter().copied());
+
+        Ok(vec![
+            null_buffer,
+            offsets_buffer,
+            empty_buffer,
+            bytes_buffer,
+        ])
+    }
+
+    fn num_buffers(&self) -> u32 {
+        self.items_decoder.num_buffers() + 2
+    }
+}
+
+#[derive(Debug)]
+pub struct DictionaryEncoder {
+    indices_encoder: Box<dyn ArrayEncoder>,
+    items_encoder: Box<dyn ArrayEncoder>,
+}
+
+impl DictionaryEncoder {
+    pub fn new(
+        indices_encoder: Box<dyn ArrayEncoder>,
+        items_encoder: Box<dyn ArrayEncoder>,
+    ) -> Self {
+        Self {
+            indices_encoder,
+            items_encoder,
+        }
+    }
+}
+
+fn get_indices_items_from_arrays(arrays: &[ArrayRef]) -> (ArrayRef, ArrayRef) {
+    let mut arr_hashmap: HashMap<&str, u8> = HashMap::new();
+    let mut curr_dict_index = 1;
+    let total_capacity = arrays.iter().map(|arr| arr.len()).sum();
+
+    let mut dict_indices = vec![0; total_capacity];
+    let mut indices_ctr = 0;
+    let mut dict_elements = Vec::new();
+
+    for arr in arrays.iter() {
+        let string_array = arrow_array::cast::as_string_array(arr);
+
+        for i in 0..string_array.len() {
+            if !string_array.is_valid(i) {
+                indices_ctr += 1;
+                continue;
+            }
+
+            let st = string_array.value(i);
+
+            let hashmap_entry = *arr_hashmap.entry(st).or_insert(curr_dict_index);
+            dict_indices[indices_ctr] = hashmap_entry;
+
+            // if item didn't exist in the hashmap, add it to the dictionary
+            // and increment the dictionary index
+            if hashmap_entry == curr_dict_index {
+                dict_elements.push(st);
+                curr_dict_index += 1;
+            }
+
+            indices_ctr += 1;
+        }
+    }
+
+    let array_dict_indices = Arc::new(UInt8Array::from(dict_indices)) as ArrayRef;
+
+    // If there is an empty dictionary:
+    // Either there is an array of nulls or an empty array altogether
+    // In this case create the dictionary with a single null element
+    // Because decoding [] is not currently supported by the binary decoder
+    // We will remove this null element later during decoding
+    let array_dict_elements: ArrayRef = if dict_elements.is_empty() {
+        Arc::new(StringArray::from(vec![Option::<&str>::None]))
+    } else {
+        Arc::new(StringArray::from(dict_elements)) as ArrayRef
```

Comment:
This is an extra copy that could be avoided by using `StringBuilder` instead of `Vec` for `dict_elements`.

That being said, the dictionary is probably pretty small,  so not a major issue.

Reply:
Fixed